usr.bin/biff/biff.1
.\"	$OpenBSD: src/usr.bin/biff/biff.1,v 1.18 2013/08/12 14:19:53 jmc Exp $
.\"	$NetBSD: biff.1,v 1.3 1995/03/26 02:34:21 glass Exp $
.\"
.\" Copyright (c) 1980, 1990, 1993
.\"	The Regents of the University of California.  All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 3. Neither the name of the University nor the names of its contributors
.\"    may be used to endorse or promote products derived from this software
.\"    without specific prior written permission.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.\"     @(#)biff.1	8.1 (Berkeley) 6/6/93
.\"
.Dd $Mdocdate: April 28 2013 $
.Dt BIFF 1
.Os
.Sh NAME
.Nm biff
.Nd be notified if mail arrives and who it is from
.Sh SYNOPSIS
.Nm biff
.Op Cm n | y
.Sh DESCRIPTION
.Nm
informs the system whether you want to be notified when mail arrives
during the current terminal session.
.Pp
The options are as follows:
.Bl -tag -width Ds
.It Cm n
Disables notification.
.It Cm y
Enables notification.
.El
.Pp
When mail notification is enabled, the header and first few lines of
the message will be printed on your screen whenever mail arrives.
A
.Dq Li biff y
command is often included in the file
.Pa \&.login
or
.Pa \&.profile
to be executed at each login.
.Pp
.Nm
operates asynchronously.
For synchronous notification use the
.Ev MAIL
variable of
.Xr sh 1
or the
.Ev mail
variable of
.Xr csh 1 .
.Sh EXIT STATUS
The
.Nm
utility exits with one of the following values:
.Bl -tag -width indent
.It 0
Notification was enabled at the time of invocation.
.It 1
Notification was disabled at the time of invocation.
.It >1
An error occurred.
.El
.Sh SEE ALSO
.Xr csh 1 ,
.Xr mail 1 ,
.Xr sh 1 ,
.Xr comsat 8
.Sh HISTORY
The
.Nm
command appeared in
.Bx 4.0 .
.Qq Biff
was Heidi Stettner's dog.
He died in August 1993, at 15.


usr.bin/biff/biff.c
/*	$OpenBSD: src/usr.bin/biff/biff.c,v 1.10 2007/02/20 13:41:19 jmc Exp $	*/
/*	$NetBSD: biff.c,v 1.3 1995/03/26 02:34:22 glass Exp $	*/

/*
 * Copyright (c) 1980, 1993
 *	The Regents of the University of California.  All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. Neither the name of the University nor the names of its contributors
 *    may be used to endorse or promote products derived from this software
 *    without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 */

#ifndef lint
static char copyright[] =
"@(#) Copyright (c) 1980, 1993\n\
	The Regents of the University of California.  All rights reserved.\n";
#endif /* not lint */

#ifndef lint
#if 0
static char sccsid[] = "@(#)biff.c	8.1 (Berkeley) 6/6/93";
#else
static char rcsid[] = "$OpenBSD: src/usr.bin/biff/biff.c,v 1.10 2007/02/20 13:41:19 jmc Exp $";
#endif
#endif /* not lint */

#include <sys/types.h>
#include <sys/stat.h>

#include <err.h>
#include <errno.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>

static void usage(void);

int
main(int argc, char *argv[])
{
	struct stat sb;
	int ch;
	char *name;


	while ((ch = getopt(argc, argv, "")) != -1)
		switch(ch) {
		case '?':
		default:
			usage();
		}
	argc -= optind;
	argv += optind;

	if ((name = ttyname(STDERR_FILENO)) == NULL)
		err(2, "tty");

	if (stat(name, &sb))
		err(2, "stat");

	if (*argv == NULL) {
		(void)printf("is %s\n", sb.st_mode & S_IXUSR ? "y" : "n");
		exit(sb.st_mode & S_IXUSR ? 0 : 1);
	}

	switch(argv[0][0]) {
	case 'n':
		if (chmod(name, sb.st_mode & ~S_IXUSR) < 0)
			err(2, "%s", name);
		break;
	case 'y':
		if (chmod(name, sb.st_mode | S_IXUSR) < 0)
			err(2, "%s", name);
		break;
	default:
		usage();
	}
	exit(sb.st_mode & S_IXUSR ? 0 : 1);
}

static void
usage(void)
{
	(void)fprintf(stderr, "usage: biff [n | y]\n");
	exit(2);
}
